### DISCUSSION TITLE
error in routing when reloading page

### DISCUSSION THREAD

**Discussion by dbehterev** [quasarframework/quasar#10941] [CLI - SSR mode]
Hello there, the second day I try to solve problem, I'll be appreciate for any help.
I have 2 categories of routes: public and secured:
router/routes.js:
```const routes = [
{
      path: '/',
      component: () => import('layouts/MainLayout.vue'),
      children: [
        {
          path: '',
          name: 'index',
          component: () => import('pages/Index.vue'),
          meta: { auth: false }
        }
     ]
...
},
{
      path: '/profile',
      meta: {
        auth: true
      },
      component: () => import('layouts/ProfileLayout.vue'),
      children: [
        {
          path: '',
          name: 'profile.index',
          component: () => import('pages/profile/Index.vue'),
          meta: {
            auth: true
          }
        }
     ]
}
```
in router/index.js I have beforeEach function:
```
context.router.beforeEach(async (to, from, next) => {
   let token = store.getters['users/tokenGetter']
  if (token) {
    return next()
  } else {
    return next(!to.meta.auth & !to.meta.verify || { path: '/login', query: { from: to.name } })
  }
})
```
So, when user go to /profile and because /profile path has meta.auth === true, user will be redirected to login page and after successful login user will be redirected on profile page. But if page will be reloaded I'll get blank page with error 'TypeError: Cannot read properties of undefined'. I noticed that problem on server side and if I put this check:
```
if (process.env.SERVER) {
  return next()
}
```
routing partially works as expected but some side effects. I don't understand why I have 2 parallel routing process - one on client side and another on server side. Client side knows everything about current user and server side knows nothing about existing user - and send routing to wrong path.. So another question - is it possible to interchange data between client and server side so when routing on server side I could send it on right path.

**Comment by suleimanas** [comment]
whats store.getters['users/tokenGetter']? from where did you get the token

**Reply by dbehterev** [reply]
> whats store.getters['users/tokenGetter']? from where did you get the token
it just return the token from state
```
export function tokenGetter (state) {
  return state.token
}
```

**Comment by hawkeye64** [comment]
@dbehterev You probably shouldn't be using the store to keep your user's authentication. Use a cookie or LocalStorage. After getting the token, you should also be checking if it is expired. So, `(no token or (token and token expired)) redirect to login page`
With this, even if the page is refreshed, it will still work.

**Reply by dbehterev** [reply]
> @dbehterev Use a cookie or LocalStorage
actually I use SessionStorage to get a token in case it persists there. I use beforeEach to extract user and token:
```
  context.router.beforeEach(async (to, from, next) => {
    if (process.env.SERVER) {
      if (/\/profile\//.test(to.path)) {
        return next()
      }
    }

    const { store } = context
    let token = store.getters['users/tokenGetter']

    const authSessionStorage = SessionStorage.getItem('auth')
    // если токен не задан в хранилище, но он есть в хранилище сессии, то перепропишем токен в локальном хранилище
    if (!token && authSessionStorage && authSessionStorage.token) {
      token = authSessionStorage.token
      store.commit('users/loginMutation', { token: authSessionStorage.token })
    }
    const user = store.getters['users/usersGetter']
    if (!user && authSessionStorage && authSessionStorage.user) {
      store.commit('users/userMutation', { user: authSessionStorage.user })
    }

    if (token) { // Authenticated Router
      return next()
    } else { // Unauthenticated Route
      return next(!to.meta.auth & !to.meta.verify || { path: '/login', query: { from: to.name } })
    } 
  })
```

**Reply by dbehterev** [reply]
So I noticed in case user on secured part of the site (for ex. /profile/smth ) and reloads a page, the whole routing process stucks. Routing process goes into two different paths:
- when it goes on server side so there no SessionStorage and server knows nothing about saved token/user and routes request to /login page and simultaneously goes another process on client side:
- but client extract saved beforehand token and user so all routing process ends up with next() function
So that's why I put check like that:
```if (process.env.SERVER) {
      if (/\/profile\//.test(to.path)) {
        return next()
      }
    }

```

**Comment by suleimanas** [comment]
https://quasar.dev/quasar-plugins/cookies

**Reply by dbehterev** [reply]
Is it possible being on client side write smth. in the cookie on server side?
I think it is common thing that is not well explained. I mean this typical scenario:
1. User has been authorized on client side and got token from back-end 
2. Then this token has been written in local state and LocalStorage/SessionStorage - this is going on client side. Is it possible at this point also write token in Cookie on server side? 
There is example how to use Cookie, but I don't understand how to get ssrContext:
```
function (**ssrContext**) {
  const cookies = process.env.SERVER
    ? Cookies.parseSSR(ssrContext)
    : Cookies // otherwise we're on client
}
```

**Reply by suleimanas** [reply]
from your backend you need set the cookie while returning the token, then you can use both in beforeEach if is't onserver use cookie or local/sessionStorage (if you need both)

**Reply by dbehterev** [reply]
> from your backend you need set the cookie while returning the token, then you can use both in beforeEach if is't onserver use cookie or local/sessionStorage (if you need both)

But that is problem - all you get on client side is not accessible on server side and vice versa. 
For example, I have interceptors in boot/axios like that:
```
export default ({ app, router, store, ssrContext }) => {
  
  // Response interceptor
  axiosInstance.interceptors.response.use(response => {
    console.log(
      '%c[RESPONSE] received -> ' + response.config.url,
      'color: green; font-weight: bold;',
      response
    )
    
    if (process.env.SERVER) {
		// never executed
         console.dir(response.data)
	}

    return response
  }, error => {
    ...
  })
}
```
On the server I don't have access to data that backend returned.

I searched another solution: persistedState. This also doesn't work. Common approach is to put into boot process code like that:
```
import { Cookies } from 'quasar'
import createPersistedState from 'vuex-persistedstate'

const persistState = function ({ name, store, storage }) {
  createPersistedState({
    key: name,
    paths: [name],
    filter ({ type }) {
      return type.startsWith(name)
    },
    storage
  })(store)
}

export default async function ({ store, ssrContext }) {
  const cookies = process.env.SERVER
    ? Cookies.parseSSR(ssrContext)
    : Cookies

  const cookieStorage = {
    getItem (key) {
      return JSON.stringify(cookies.get(key))
    },
    setItem (key, value) {
      cookies.set(key, value, { path: '/' })
    },
    removeItem (key) {
      cookies.remove(key, { path: '/' })
    }
  }

  persistState({ name: 'app', store, storage: cookieStorage })

  if (process.env.CLIENT) {
    persistState({ name: 'local', store, storage: window.localStorage })
  }
}
```
But all I see that strange data set into the cookie:
```
Set-Cookie: @@=1; Path=/
```
I also tried to put into plugins section of Vuex.Store persistedState, but it throws error:  UnhandledPromiseRejectionWarning: ReferenceError: window is not defined
I think because on node side the code is trying to access 'window' that exists only in client environment. 
```
import createPersistedState from 'vuex-persistedstate'

export default function ({ store, ssrContext }) {
  const Store = new Vuex.Store({
    modules: {
      users,
      config,
      servers,
      routes,
      cart
    },
    strict: process.env.DEV
    plugins: [createPersistedState()]
  })

  return Store
}
```
